src/Storages/Distributed/DirectoryMonitor.cpp

```diff
@@ -502,16 +504,16 @@ std::map<UInt64, std::string> StorageDistributedDirectoryMonitor::getFiles()
     }
 
     {
-        std::lock_guard metrics_lock(metrics_mutex);
+        std::lock_guard status_lock(status_mutex);
 
-        if (files_count != files.size())
-            LOG_TRACE(log, "Files set to {} (was {})", files.size(), files_count);
-        if (bytes_count != new_bytes_count)
-            LOG_TRACE(log, "Bytes set to {} (was {})", new_bytes_count, bytes_count);
+        if (status.files_count != files.size())
+            LOG_TRACE(log, "Files set to {} (was {})", files.size(), status.files_count);
+        if (status.bytes_count != new_bytes_count)
+            LOG_TRACE(log, "Bytes set to {} (was {})", new_bytes_count, status.bytes_count);
 
         metric_pending_files.changeTo(files.size());
-        files_count = files.size();
-        bytes_count = new_bytes_count;
+        status.files_count = files.size();
+        status.bytes_count = new_bytes_count;
```

Comment:
Here's a fundamental race condition. See https://pastila.nl/?00950e00/41a3c7bbb0a7e75bd3f2922c58b02334
The problem is that INSERT have already created some file on disk, but have not incremented `files_count` yet.
So `getFiles()` counted this file and decided to update `files_count` from 0 to 1.
And then INSERT incremented it too, so it became 2.

Why do we need this async `status` update at all? Can we just count files on startup and then maintain correct status by incrementing/decrementing the counters for each insert/flush? Even if we can not, maybe we should remove all increments/decrements of the counters and keep only async update. Because currently it's a bit misleading.

Reply:
Nice catch!

Instant metrics updates is required for more real-time data in system.distribution_queue.
And "async" (here it is in quotes because it is racy only with INSERT that schedules the file and the monitor), is to keep it 100% accurate.

I've looked at production at got this:

```
Bytes set to 20941765 (was 21655447)
Bytes set to 713682 (was 0)
Bytes set to 33041559 (was 35609256)
Bytes set to 2567697 (was 0)
```

So `getFiles` sometimes does updates the counters, and I don't have other messages (that has negative values in the counter now, that will show that indeed some race had happened) like in provided logs, or maybe I'm misinterpreting something.

This can be fixed with a lock, but it is a bad idea to hold it for the whole dir scan time.

Will take a deeper look.

Reply:
Actually everything is updated correctly without async update, so let's just remove it - #43406